h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/exec/RemoteXGBoostExecutor.java

```diff
@@ -0,0 +1,123 @@
+import ml.dmlc.xgboost4j.java.XGBoostMatrixFactory;
+import ml.dmlc.xgboost4j.java.XGBoostSaveMatrixTask;
+import ml.dmlc.xgboost4j.java.XGBoostSetupTask;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.util.EntityUtils;
+import water.AutoBuffer;
+import water.H2O;
+import water.Key;
+import water.fvec.Frame;
+import water.util.IcedHashMap;
+import water.util.Log;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static hex.schemas.XGBoostExecReqV3.Type.*;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.http.HttpHeaders.CONTENT_TYPE;
+import static water.util.HttpResponseStatus.OK;
+
+public class RemoteXGBoostExecutor implements XGBoostExecutor {
+
+    public final String uri;
+    public final Key modelKey;
+
+    public RemoteXGBoostExecutor(String remoteUri, XGBoostModel model, Frame train, XGBoostModel.XGBoostParameters parms) {
+        uri = remoteUri + "/3/XGBoostExecutor";
+        modelKey = model._key;
+        XGBoostExecReq.Init req = new XGBoostExecReq.Init();
+        XGBoostSetupTask.FrameNodes trainFrameNodes = XGBoostSetupTask.findFrameNodes(train);
+        req.num_nodes = trainFrameNodes.getNumNodes();
+        if (parms.hasCheckpoint()) {
+            req.checkpoint_bytes = model.model_info()._boosterBytes;
+        }
+        DataInfo dataInfo = model.model_info().dataInfo();
+        req.parms = new IcedHashMap<>();
+        req.parms.putAll(XGBoostModel.createParamsMap(parms, model._output.nclasses(), dataInfo.coefNames()));
+        model._output._native_parameters = BoosterParms.fromMap(req.parms).toTwoDimTable();
+        XGBoostMatrixFactory f = new FrameXGBoostMatrixFactory(model, parms, trainFrameNodes);
+        req.matrix_dir_path = H2O.ICE_ROOT.toString() + "/" + modelKey.toString();
+        new XGBoostSaveMatrixTask(modelKey, req.matrix_dir_path, trainFrameNodes._nodes, f).run();
+        req.featureMap = XGBoostUtils.createFeatureMap(model, train);
+        XGBoostExecRespV3 resp = post(modelKey, XGBoostExecReqV3.Type.INIT, req);
+        assert modelKey.equals(resp.key.key());
+    }
+    
+    private XGBoostExecRespV3 post(Key key, XGBoostExecReqV3.Type type, XGBoostExecReq reqContent) {
+        XGBoostExecReqV3 req = new XGBoostExecReqV3(key, type, reqContent);
+        HttpPost httpReq = new HttpPost(uri);
+        httpReq.setEntity(new StringEntity(req.toJsonString(), UTF_8));
+        httpReq.setHeader(CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType());
+        Log.info("Remote XGBoost: Request " + type + " " + reqContent);
+        try (CloseableHttpClient client = HttpClientBuilder.create().build();
```

Comment:
Connection pooling is not essential to us, as we don't stand and fall with REST API calls performance. It's worth thinking of it. Apache HTTP has connection pooling built-in. Moving this to another layer with reasonable connection pool timeouts (all the way down to absolute zero) sounds correct

https://hc.apache.org/httpcomponents-client-ga/tutorial/html/connmgmt.html

If H2O had JAX-RS or similar, with some reasonable implementation like Jersey (https://eclipse-ee4j.github.io/jersey/), our life would be easier.

Reply:
thanks will add to the list of things to do in the final impl